nixos/tpm2, tpm2-tss, tpm2-pytss: improve usability of TPM2 FAPI

scottstephens opened the pull request:
Using the TPM2 FAPI, such as through the tss2 cli, was inconvenient in many ways. For instance, in the tss2 commands, the path to the default fapi-config file was specified at compile time to a path in the nix store. That default config was also unusable, as it pointed logging and other mutable directories to paths in the nix store. Deploying a change of security.tpm2.enable from false to true would also generally fail on the first try, due to lacking a udev trigger to cause the new udev rules to actually run.

Along the way I cleaned up some other issues I found. The individual commits have the details on the specific changes and the reasons for them.

Fixes #192771.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [x] [NixOS tests] in [nixos/tests].
  - [x] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [ ] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

scottstephens committed:
tpm2-tss: add FAPI config for testing

scottstephens committed:
tpm2-pytss: allow using EV in tests, use EV

There was a bug in upstream that prevented environment
variables from being used to configure FAPI for testing.

This incorporates a fix that is now in upstream but not yet
released, and then switches to using environment variables
instead of patching to configure the tests.

scottstephens committed:
tpm2-tss: change default path of fapi config file

The default path is hard-coded at compile time. Previously it pointed to
a location inside the nix store, which made it necessary to use
environment variables to specify the path to a user-specified file.

The nature of the config file is such that it's not possible to specify
one that works for every system this package may be installed in, so
user-configuration is usually necessary.

The default is now /etc/tpm2-tss/fapi-config.json, which
is more compatible with non-nixos config conventions. NixOS
compatibility is now accomplished by linking from the well-known
location to a config file in the nix store.

scottstephens committed:
nixos/tpm2: add fapi options section

This allows a user to generate a fapi-config.json file and links
/etc/tpm2-tss/fapi-config.json to it.

scottstephens committed:
nixos/tpm2: trigger udev change on tpm devices

The package changes udev rules. A device change must be triggered
in order for these changes to be picked up, otherwise services
which depend on the configuration will fail.

scottstephens committed:
nixos/tpm2: create directories for FAPI

scottstephens committed:
nixos/tpm2: add tests

scottstephens committed:
nixos/tpm2: start tpm2-abrmd after dev-tpm0.device

tpm2-abrmd can't start up until /dev/tpm0 has been created and its
udev rules applied. This prevents a failure on startup by waiting
for that to happen.

scottstephens committed:
nixos/tpm2: add systemd tag to udev rules

Updating the udev rules to match what is in recent versions of the
suggested rules from the tpm2-tss package. The old rules were based
on an old version of the suggested rules and never updated when
they were updated in subsequent version upgrades.

scottstephens commented:
@baloo @lschuermann I think you may be appropriate reviewers for this. My apologies if pinging you is premature; I don't understand what's going on with the CI very well, but it looks to me like it needs intervention from a maintainer to progress any further.

scottstephens commented:
@acid-bong done

philiptaron reviewed (commented):
Thanks for your work here!

scottstephens committed:
nixos/tpm2: use redirection instead of cat

Co-authored-by: Philip Taron <philip.taron@gmail.com>

philiptaron reviewed (approved):
I'm approving this and will merge after:

1. Targeting staging (see my comment on how to get the PR's base set up for still running tests.)
2. Adding the excellent explanations you wrote for me into the code and descriptions themselves. I'll trust your judgement on where they ought to land.

scottstephens committed:
nixos/tpm2: improve module options documentation

scottstephens committed:
nixos/tpm2: add comment describing purpose of script

scottstephens committed:
nixos/tpm2: use better method for udev trigger

scottstephens committed:
nixos/tpm2: add manual entry

scottstephens commented:
I changed the base branch, spruced up the documentation, and changed the method for triggering udev changes.

The bulk of the new documentation is in a new page in the NixOS manual that describes the basics of what TPM hardware does and how the software ecosystem for it is organized.

philiptaron commented:
I intend to review and merge this sometime in the next couple days

philiptaron commented:
Yeah, don't worry about it.

philiptaron commented:
> What test caught that shell script problem? That would be handy to run locally.

I checked out the branch locally and ran `nom-build -A nixosTests.tpm2`

philiptaron commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 440251 --checkout commit --package nixosTests.tpm2`
Commit: `0509fb6f9764f44f250496d1636fe8cfcf684028`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 1 test built:</summary>
  <ul>
    <li>nixosTests.tpm2</li>
  </ul>
</details>
